File: .skills/openclaw-skills/skills/ogenes/dingtalk-api/scripts/create-approval-instance.ts
// 发起审批实例
// 用法: ts-node scripts/create-approval-instance.ts <processCode> <originatorUserId> <deptId> <formValuesJson> [--ccList "user1,user2"] [--debug]
// 需要设置环境变量 DINGTALK_APP_KEY 和 DINGTALK_APP_SECRET
export {};

const { default: dingtalkWorkflow, CreateProcessInstanceHeaders, CreateProcessInstanceRequest } = require('@alicloud/dingtalk/workflow_1_0');
const { default: dingtalkOauth2_1_0, GetAccessTokenRequest } = require('@alicloud/dingtalk/oauth2_1_0');
const { Config } = require('@alicloud/openapi-client');
const { RuntimeOptions } = require('@alicloud/tea-util');

interface Result {
  success: boolean;
  processCode: string;
  originatorUserId: string;
  instanceId: string;
}

interface ErrorResult {
  success: false;
  error: {
    code: string;
    message: string;
  };
}

function createConfig(): any {
  const config = new Config({});
  config.protocol = "https";
  config.regionId = "central";
  return config;
}

async function getAccessToken(appKey: string, appSecret: string): Promise<string> {
  const config = createConfig();
  const client = new dingtalkOauth2_1_0(config);

  const request = new GetAccessTokenRequest({
    appKey: appKey,
    appSecret: appSecret,
  });

  try {
    const response = await client.getAccessToken(request);
    const accessToken = response.body?.accessToken;

    if (!accessToken) {
      throw new Error('获取 access_token 失败：响应中未包含 token');
    }

    return accessToken;
  } catch (err: any) {
    throw new Error(`获取 access_token 失败: ${err.message || err}`);
  }
}

async function createApprovalInstance(
  accessToken: string,
  processCode: string,
  originatorUserId: string,
  deptId: string,
  formComponentValues: any[],
  ccList: string | undefined,
  debug: boolean
): Promise<void> {
  const client = new dingtalkWorkflow(createConfig());

  const headers = new CreateProcessInstanceHeaders({});
  headers.xAcsDingtalkAccessToken = accessToken;

  const request = new CreateProcessInstanceRequest({
    processCode: processCode,
    originatorUserId: originatorUserId,
    deptId: deptId,
    formComponentValues: formComponentValues,
    ccList: ccList,
  });

  try {
    const response = await client.createProcessInstanceWithOptions(
      request,
      headers,
      new RuntimeOptions({})
    );

    if (debug) {
      console.error('\n=== 调试信息 ===');
      console.error('完整响应:', JSON.stringify(response, null, 2));
      console.error('==============\n');
    }

    const result: Result = {
      success: true,
      processCode: processCode,
      originatorUserId: originatorUserId,
      instanceId: response.body?.result?.processInstanceId || '',
    };

    console.log(JSON.stringify(result, null, 2));
  } catch (err: any) {
    const errorResult: ErrorResult = {
      success: false,
      error: {
        code: err.code || 'UNKNOWN_ERROR',
        message: err.message || '未知错误',
      }
    };
    console.error(JSON.stringify(errorResult, null, 2));
    process.exit(1);
  }
}

function parseArgs(args: string[]): { processCode: string; originatorUserId: string; deptId: string; formValuesJson: string; ccList?: string; debug: boolean } {
  let processCode = '';
  let originatorUserId = '';
  let deptId = '';
  let formValuesJson = '';
  let ccList: string | undefined;
  let debug = false;

  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '--debug') {
      debug = true;
    } else if (arg === '--ccList' && i + 1 < args.length) {
      ccList = args[++i];
    } else if (!arg.startsWith('--')) {
      if (!processCode) {
        processCode = arg;
      } else if (!originatorUserId) {
        originatorUserId = arg;
      } else if (!deptId) {
        deptId = arg;
      } else if (!formValuesJson) {
        formValuesJson = arg;
      }
    }
  }

  return { processCode, originatorUserId, deptId, formValuesJson, ccList, debug };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const { processCode, originatorUserId, deptId, formValuesJson, ccList, debug } = parseArgs(args);

  const appKey = process.env.DINGTALK_APP_KEY;
  const appSecret = process.env.DINGTALK_APP_SECRET;

  if (!appKey || !appSecret) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'MISSING_CREDENTIALS',
        message: '缺少钉钉应用凭证，请设置环境变量 DINGTALK_APP_KEY 和 DINGTALK_APP_SECRET',
        usage: 'export DINGTALK_APP_KEY=your-app-key && export DINGTALK_APP_SECRET=your-app-secret && ts-node scripts/create-approval-instance.ts <processCode> <originatorUserId> <deptId> \'[{"name":"标题","value":"请假申请"}]\' [--ccList "user1,user2"] [--debug]'
      }
    }, null, 2));
    process.exit(1);
  }

  if (!processCode || !originatorUserId || !deptId || !formValuesJson) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'INVALID_ARGUMENTS',
        message: '参数错误：需要提供 processCode, originatorUserId, deptId 和 formValuesJson',
        usage: 'ts-node scripts/create-approval-instance.ts <processCode> <originatorUserId> <deptId> \'[{"name":"标题","value":"请假申请"}]\' [--ccList "user1,user2"] [--debug]'
      }
    }, null, 2));
    process.exit(1);
  }

  let formComponentValues: any[];
  try {
    formComponentValues = JSON.parse(formValuesJson);
  } catch (e) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'INVALID_JSON',
        message: 'formValuesJson 参数不是有效的 JSON 字符串',
      }
    }, null, 2));
    process.exit(1);
  }

  try {
    console.error('正在获取 access_token...');
    const accessToken = await getAccessToken(appKey, appSecret);
    console.error('access_token 获取成功，正在发起审批实例...');

    await createApprovalInstance(accessToken, processCode, originatorUserId, deptId, formComponentValues, ccList, debug);
  } catch (err: any) {
    console.error(JSON.stringify({
      success: false,
      error: {
        code: 'AUTH_FAILED',
        message: err.message || '认证失败',
      }
    }, null, 2));
    process.exit(1);
  }
}

main();

File: .skills/openclaw-skills/skills/hioneowner/ding-skills/scripts/create_approval_instance.py
"""发起审批实例

用法: python scripts/create_approval_instance.py "<processCode>" "<originatorUserId>" "<deptId>" '<formValuesJson>' [--ccList "user1,user2"]

formValuesJson 示例: '[{"name":"标题","value":"请假申请"}]'
"""

import sys
import os
import json as json_mod
sys.path.insert(0, os.path.dirname(__file__))
from dingtalk_client import get_access_token, api_request, handle_error, output


def parse_args(argv):
    args = {"process_code": None, "originator_user_id": None, "dept_id": None, "form_values_json": None, "cc_list": None}
    positional = []
    i = 1
    while i < len(argv):
        if argv[i] == "--ccList" and i + 1 < len(argv):
            args["cc_list"] = argv[i + 1]
            i += 2
        elif argv[i] == "--debug":
            i += 1
        elif not argv[i].startswith("--"):
            positional.append(argv[i])
            i += 1
        else:
            i += 1
    if len(positional) >= 1:
        args["process_code"] = positional[0]
    if len(positional) >= 2:
        args["originator_user_id"] = positional[1]
    if len(positional) >= 3:
        args["dept_id"] = positional[2]
    if len(positional) >= 4:
        args["form_values_json"] = positional[3]
    return args


def main():
    args = parse_args(sys.argv)
    if not all([args["process_code"], args["originator_user_id"], args["dept_id"], args["form_values_json"]]):
        output({"success": False, "error": {"code": "INVALID_ARGS",
            "message": "用法: python scripts/create_approval_instance.py \"<processCode>\" \"<originatorUserId>\" \"<deptId>\" '<formValuesJson>' [--ccList \"user1,user2\"]"}})
        sys.exit(1)

    try:
        form_values = json_mod.loads(args["form_values_json"])
    except json_mod.JSONDecodeError:
        output({"success": False, "error": {"code": "INVALID_JSON", "message": "formValuesJson 参数不是有效的 JSON 字符串"}})
        sys.exit(1)

    try:
        token = get_access_token()
        print("正在发起审批实例...", file=sys.stderr)

        body = {
            "processCode": args["process_code"],
            "originatorUserId": args["originator_user_id"],
            "deptId": args["dept_id"],
            "formComponentValues": form_values,
        }
        if args["cc_list"]:
            body["ccList"] = args["cc_list"]

        result = api_request("POST", "/workflow/processInstances", token, json_body=body)
        output({
            "success": True,
            "processCode": args["process_code"],
            "originatorUserId": args["originator_user_id"],
            "instanceId": result.get("result", {}).get("processInstanceId", ""),
        })
    except Exception as e:
        handle_error(e)
        sys.exit(1)


if __name__ == "__main__":
    main()
